### PR TITLE
Improvements for deploying OpenEBS on AWS Cloud Platform.

### PR DESCRIPTION
**Pull request by yudaykiran**
Improvements for deploying OpenEBS on AWS Cloud Platform.

Code Changes:
-------------------
- Add support for CoreOS.
- Add option for listing EC2 instances created on AWS.
- Automatically deploy OpenEBS on AWS Cloud.
- Add validations for IPAddress and AWS EC2 Instances.
- Add support for user defined <username> in AWS IAM.

Documentation Changes:
--------------------------------
- Update docs with latest changes to script.

Tested On:
-------------
Amazon Web Services with Ubuntu 16.04.

Details of the fix:
--------------------
The fix contains enhancements for deploying OpenEBS on Amazon Cloud Platform. Support for CoreOS has been implemented. The fix also improves validations performed for IPAddresses and AWS EC2 instances. 

Signed-off-by: yudaykiran <udaykiran.y@gmail.com>

### CODE REVIEW

File: e2e/terraform/README.md [hunk 1]
```diff
@@ -17,21 +17,21 @@ People who already have an AWS account can skip the above step.
 - Select *IAM* under *Security, Identity & Compliance*.
 - In the *Dashboard*, click *Users*.
 - Click *Add User* button.
-- In the *User name* textbox, enter *openebsuser01* as the username.
+- In the *User name* textbox, type the name of the user you want to create. For Example: *openebsuser*
```
Comment (mahebbar) [inline e2e/terraform/README.md]: enter the name of the user you want to create. For example, *openebuser*

Comment (yudaykiran) [reply e2e/terraform/README.md]: @mahebbar, thanks. The new commit contains the changes as suggested.

**Comment by kmova** [comment]
hi @yudaykiran, can you submit the reports / outputs of running the scripts with different omi's.

**Comment by yudaykiran** [comment]
Made changes based on review comments. Kindly review the new changes.

**Comment by ksatchit** [comment]
Good additions for starting iscsi services !!

File: e2e/terraform/README.md [hunk 2]
```diff
@@ -147,26 +150,55 @@ Common commands:
...
+**For CoreOS**
+
+```
+$ ./oebs-cloud.sh --ssh-aws-ec2
+Container Linux by CoreOS stable (1465.6.0)
+core@ip-172-20-53-140 ~ $
+```
+
+- Running `--ssh-aws-ec2` without any arguments, by default connects you to the Kubernetes Master.
+    - `--ssh-aws-ec2` can also be run as `--ssh-aws-ec2=ipaddress`, where `ipaddress` is the Public IPAddress of the AWS EC2 instance.
```
Comment (mahebbar) [inline e2e/terraform/README.md]: You can also run `--ssh-aws-ec2` as.......IP (space)

File: e2e/terraform/README.md [hunk 3]
```diff
@@ -147,26 +150,55 @@ Common commands:
   - Fix these errors and re-verify with `terraform plan` before running the `terraform apply` command.
 - Run the command `terraform apply` to initiate creation of the infrastructure.
```
Comment (mahebbar) [inline e2e/terraform/README.md]: to initiate infrastructure recreation

File: e2e/terraform/README.md [hunk 4]
```diff
@@ -147,26 +150,55 @@ Common commands:
   - Fix these errors and re-verify with `terraform plan` before running the `terraform apply` command.
```
Comment (mahebbar) [inline e2e/terraform/README.md]: re-verify = verify again?

File: e2e/terraform/README.md [hunk 5]
```diff
@@ -111,6 +112,8 @@ $ source ~/.profile
 $ ./oebs-cloud.sh --create-cluster-config
 ```
 
+- Running `--create-cluster-config` without any arguments defaults to `Ubuntu`.
+  - `--create-cluster-config` can be run with `--ami-vm-os=ubuntu` or `--ami-vm-os=coreos`.
 - A terraform file `kubernetes.tf` is generated in the same directory.
 - Passwordless SSH connection between the local workstation and the remote EC2 instances is established.
```
Comment (mahebbar) [inline e2e/terraform/README.md]: SSH connection without a password

File: e2e/terraform/README.md [hunk 6]
```diff
@@ -102,6 +102,7 @@ $ source ~/.profile
 ```
 
 ## Creating the Config For Cluster
+
 - We will be generating a terraform file(.tf) that will later spawn:
```
Comment (mahebbar) [inline e2e/terraform/README.md]: Who is we? OpenEBS? 
...that will spawn (delete later)

File: e2e/terraform/README.md [hunk 7]
```diff
@@ -102,6 +102,7 @@ $ source ~/.profile
 ```
 
 ## Creating the Config For Cluster
```
Comment (mahebbar) [inline e2e/terraform/README.md]: Configuring Cluster or is it Creating the Configuration File for Cluster?

File: e2e/terraform/README.md [hunk 8]
```diff
@@ -147,26 +150,55 @@ Common commands:
...
+core@ip-172-20-53-140 ~ $
+```
+
+- Running `--ssh-aws-ec2` without any arguments, by default connects you to the Kubernetes Master.
+    - `--ssh-aws-ec2` can also be run as `--ssh-aws-ec2=ipaddress`, where `ipaddress` is the Public IPAddress of the AWS EC2 instance.
+- You should now be running inside the AWS EC2 instance.
 
 ## Deploy OpenEBS on AWS
 
 Deploying OpenEBS requires Kubernetes to be already running on the EC2 instances. Verify if a Kubernetes cluster has been created.
```
Comment (mahebbar) [inline e2e/terraform/README.md]: replace to be already = Kubernetes must be running on the EC2 instances while deploying OpenEBS. 
replace has been = Verify if a Kubernetes cluster is created.

File: e2e/terraform/README.md [hunk 9]
```diff
@@ -147,26 +150,55 @@ Common commands:
...
 NAME                            STATUS    AGE       VERSION 
-ip-172-20-36-126.ec2.internal   Ready     1m        v1.7.0 
-ip-172-20-37-115.ec2.internal   Ready     1m        v1.7.0 
-ip-172-20-53-140.ec2.internal   Ready     3m        v1.7.0
+ip-172-20-36-126.ec2.internal   Ready     1m        v1.7.2 
+ip-172-20-37-115.ec2.internal   Ready     1m        v1.7.2 
+ip-172-20-53-140.ec2.internal   Ready     3m        v1.7.2
 ```
 
 OpenEBS is already deployment by the time you are logged into AWS (Amazon Web Services).
```
Comment (mahebbar) [inline e2e/terraform/README.md]: OpenEBS is deployed by the time...

Comment (mahebbar) [reply e2e/terraform/README.md]: OpenEBS is deployed by the time you login to

File: e2e/terraform/README.md [hunk 10]
```diff
@@ -178,3 +210,23 @@ maya-apiserver-h714w      1/1       Running   0          12m
...
+
+```
+core@ip-172-20-53-140:~$ kubectl get nodes 
+NAME                            STATUS    AGE       VERSION 
+ip-172-20-36-126.ec2.internal   Ready     1m        v1.7.2 
+ip-172-20-37-115.ec2.internal   Ready     1m        v1.7.2 
+ip-172-20-53-140.ec2.internal   Ready     3m        v1.7.2
+```
+
+OpenEBS is already deployment by the time you are logged into AWS (Amazon Web Services).
```
Comment (mahebbar) [inline e2e/terraform/README.md]: OpenEBS is deployed by the time you login to

File: e2e/terraform/oebs-cloud.sh [hunk 11]
```diff
@@ -66,6 +66,15 @@ resource "aws_security_group_rule" "allow_all_node_outbound" {
...
+    cat <<'EOF' >>data/aws_launch_configuration_nodes.openebs.k8s.local_user_data
+sudo modprobe iscsi_tcp
+sudo systemctl start iscsid-initiatorname
+sudo systemctl start iscsid
+EOF
+}
+
 function setup_local_env() {
 
     echo "Installing Pre-requisites..."
```
Comment (mahebbar) [inline e2e/terraform/oebs-cloud.sh]: Prerequisites

File: e2e/terraform/README.md (hunk 3 above)
Comment (yudaykiran) [reply e2e/terraform/README.md]: recreation or creation? we do not have the infrastructure yet, at this point of deployment.

File: e2e/terraform/README.md (hunk 4 above)
Comment (yudaykiran) [reply e2e/terraform/README.md]: yes, verify again.

File: e2e/terraform/README.md (hunk 5 above)
Comment (yudaykiran) [reply e2e/terraform/README.md]: It's usually called Passwordless SSH.

File: e2e/terraform/README.md (hunk 6 above)
Comment (yudaykiran) [reply e2e/terraform/README.md]: I think I will change it to " The script will generate" is that fine?

File: e2e/terraform/README.md (hunk 7 above)
Comment (yudaykiran) [reply e2e/terraform/README.md]: Creating the Configuration File for Cluster

**Comment by yudaykiran** [comment]
Implemented the changes requested in the recent review comments.

File: e2e/terraform/README.md (hunk 5 above)
Comment (mahebbar) [reply e2e/terraform/README.md]: Learnt new! Tks!

File: e2e/terraform/README.md (hunk 3 above)
Comment (mahebbar) [reply e2e/terraform/README.md]: creation...my bad.

File: e2e/terraform/README.md [hunk 12]
```diff
@@ -101,9 +101,9 @@ PATH="$HOME/bin:$HOME/.local/bin:/usr/local/bin:$PATH"
 $ source ~/.profile
 ```
 
-## Creating the Config For Cluster
+## Creating the Configuration for Cluster
```
Comment (mahebbar) [inline e2e/terraform/README.md]: Creating the Configuration for Cluster = Creating the Configuration File for Cluster

**Comment by ksatchit** [comment]
LGTM :)

File: e2e/terraform/README.md (hunk 12 above)
Comment (yudaykiran) [reply e2e/terraform/README.md]: Done

**Pull request merged**